File: canada_code/normalise_rows.m
function [A_New]=normalise_rows(A_Old,Eps)

[N,~]=size(A_Old);%number of rows

A_New=A_Old;

%For negative values,shift all values by magnitude of minimum value
Count1=sum(A_Old(:) < 0);%Obtain number of negative values 
if Count1>=1
    A_New=A_Old-min(A_Old(:));
end


Count2=sum(A_New(:) == 0);%Number of elements with zeros


if Count2>=1
    A_New(A_New==0)=Eps;%Replace zero with small value
    for i=1:N   
        A_New(i,:)=A_New(i,:)/sum(A_New(i,:));
    end
end

for i=1:N   
    if sum(A_New(i,:))>1
        A_New(i,:)=A_New(i,:)/sum(A_New(i,:));
    end
end

end

File: normalise_rows.py
import numpy as np

def normalise_rows(A_Old, Eps):
    # number of rows
    N, _ = A_Old.shape

    # create a copy of A_Old
    A_New = A_Old.copy()

    # For negative values, shift all values by the magnitude of the minimum value
    if np.sum(A_Old<0) > 0:
        A_New = A_Old - np.min(A_Old)

    # Number of elements with zeros
    if np.sum(A_New == 0) > 0:
        A_New[np.where(A_New == 0)] = Eps  # replace zero with small value
        A_New = A_New / np.sum(A_New, axis=1, keepdims=True)

    # Normalize rows
    for i in range(N):
        if np.sum(A_New[i, :]) > 0:
            A_New[i, :] = A_New[i, :] / np.sum(A_New[i, :])

    return A_New
